drivers/linstorvolumemanager.py

```diff
@@ -196,17 +202,88 @@ def _get_controller_uri():
         # Not found, maybe we are trying to create the SR...
         pass
 
-def get_controller_uri():
-    retries = 0
-    while True:
-        uri = _get_controller_uri()
+
+@contextlib.contextmanager
+def shared_reader(path):
+    with open(path, 'r') as f:
+        lock = flock.ReadLock(f.fileno())
+        lock.lock()
+        try:
+            yield f
+        finally:
+            lock.unlock()
+
+
+@contextlib.contextmanager
+def excl_writer(path):
+    with open(path, 'r+') as f:
+        lock = flock.WriteLock(f.fileno())
+        lock.lock()
+        try:
+            yield f
+        finally:
+            lock.unlock()
+
+
+def get_cached_controller_uri(ctx=None):
+    try:
+        with ctx if ctx is not None else shared_reader(CONTROLLER_CACHE_PATH) as f:
```

Comment:
```suggestion
        with ctx if ctx else shared_reader(CONTROLLER_CACHE_PATH) as f:
```
Should be equivalent but easier to read

Reply:
fixed in [363372d](https://github.com/xcp-ng/sm/pull/83/commits/363372ddda9d0cdb53e1c2a29b089d0a0e188cad)